ethcontract-common/src/artifact/hardhat.rs

```diff
@@ -426,15 +426,16 @@ impl HardHatLoader {
         address: Address,
         transaction_hash: Option<TransactionHash>,
     ) -> Result<(), ArtifactError> {
-        let mut contract = match artifact.get_mut(&contract.name) {
-            Some(existing_contract) => {
-                if existing_contract.abi != contract.abi {
-                    return Err(ArtifactError::AbiMismatch(contract.name));
-                }
-
-                existing_contract
+        let contract_guard = artifact.get_mut(&contract.name);
+        let mut contract = if let Some(existing_contract) = contract_guard {
+            if existing_contract.interface != contract.interface {
+                return Err(ArtifactError::AbiMismatch(contract.name));
             }
-            None => artifact.insert(contract).inserted_contract,
+
+            existing_contract
+        } else {
+            drop(contract_guard);
```

Comment:
This contract_guard now also has to get dropped explicitly due to the guard modifying the contract on drop.

Reply:
Should probably be a comment in the code